pkg/smokescreen/smokescreen.go

```diff
@@ -414,32 +416,68 @@ func newContext(cfg *Config, proxyType string, req *http.Request) *smokescreenCo
 	}
 }
 
-// This is hacky but necessary in order to remove square brackets from non-IPv6 addresses
-// Otherwise, we'd be checking our ACL for "[stripe.com]" rather than "stripe.com"
-func normalizeHost(host, scheme string) (string, error) {
-	hostname, port, err := net.SplitHostPort(host)
-	if err != nil {
-		if strings.Contains(err.Error(), "missing port in address") {
-			// net.SplitHostPort should remove square brackets on a valid IPv6 address, so if there
-			// are still square brackets, the hostname is malformed and we should just return an error
-			if strings.HasPrefix(host, "[") && strings.HasSuffix(host, "]") {
-				return "", denyError{fmt.Errorf("unable to parse destination host")}
-			}
+// hasPort returns true if the provided address does not include a port number.
+func hasPort(s string) bool {
+	return strings.LastIndex(s, "]") < strings.LastIndex(s, ":")
+}
 
-			hostname = host
-			switch scheme {
-			case "http":
-				port = "80"
-			case "https":
-				port = "443"
-			default:
-				port = "0"
-			}
-		} else {
-			return "", err
+// normalizeHost returns host (as string) and port (as int) derived from the
+// `hostPort` string. `hostPort` is a colon-separated (':') DNS name and port.
+//
+// If no port is specified, the scheme string is used to find the default port.
+// If forceFQDN is true, returned normalized domain name will be an FQDN.
+//
+// Returns a normalized representation of host (Punycode for DNS names,
+// standardized IP address representation), as well as the port number.
+func normalizeHost(hostPort, scheme string, forceFQDN bool) (string, int, error) {
+	var err error
+	const noPort = -1
+	host, port := hostPort, noPort
+
+	// net.SplitHostPort() doesn't handle bare IPv6 addresses well so
+	// handle that case first.
+	if ip := net.ParseIP(hostPort); ip != nil && ip.To4() == nil {
+		host = ip.String()
+	} else if hasPort(hostPort) {
+		// Extract host and port if both are provided.
+		var portString string
+		host, portString, err = net.SplitHostPort(hostPort)
+		if err != nil {
+			return "", noPort, fmt.Errorf("unable to parse host: %v", err)
+		}
```

Comment:
If we reached this code `hasPort()` returned true. So either it's a bare IPv6 address (already ruled out in the opening if statement) or the `hostPort` is indeed malformed (since it was supposed to contain a port but it does not).